### ISSUE TITLE
Getting inputs from a file.

### ISSUE THREAD

**Issue by jayantbh** [docker-exec/dexec#55]
Trying to get file data this way:
```
dexec test.c < text.txt
```
Doesn't work.

What are the alternatives? Or how do we do this?

**Comment by bhskt** [comment]
+1 Need a way to pipe an input text file to stdin, instead of opening the file from the program and reading from it. Advice please.

**Comment by andystanton** [comment]
I recently added the ability for dexec to accept program source from stdin (https://github.com/docker-exec/dexec#reading-from-stdin). This allows you to do stuff like:

```sh
$ cat helloworld.c | dexec --extension=c
$ dexec --extension=c <helloworld.c
```

So dexec is capable of reading from stdin, it's just that at the moment it can only use that to generate temporary source files rather than passing input to the executing code.

I'll see if it's possible to combine these two ideas. If it is, I'll add it as a feature. If it's only possible to pipe one kind of thing, it makes sense to pipe input rather than source. Either way - I'll make it possible to do what you're asking.

**Comment by jayantbh** [comment]
Just a suggestion, probably a rather obvious one.

Maybe something like this can be done:
$ dexec helloworld.c --input myinput.txt

And behind the scenes, after compilation, the program can take that file as a piped input through stdin? Did that make sense?

**Comment by andystanton** [comment]
That would certainly work @jayantbh but I wonder if we can keep it more unix like. I wrote the following shell scripts to see how bash treats piped input and redirected files:

#### parent.sh

```sh
#!/usr/bin/env bash
read input
echo "parent: $input"
./child.sh
```

#### child.sh

```sh
#!/usr/bin/env bash
read input
echo "child: $input"
```

#### Usage

```sh
$ echo "a" | ./parent.sh <<<"b"
parent: a
child: b
```

So it looks like the convention is that the input piped from stdin go to the parent process and the input redirected from a file goes to the child process. With dexec, to handle both piped source code and redirected data, the call might look something like this:

```sh
$ curl http://url-to-some-code | dexec --extension cpp < input-data.txt
```

I'll see if I can get something like this working.

**Comment by jayantbh** [comment]
Well, if you can make this happen, there's nothing better! :smile: 
```
$ dexec helloworld.c < myinput.txt

**Comment by andystanton** [comment]
Yes sorry - that will also work, but I want to consider the scenario where source is piped in too! :)

**Comment by andystanton** [comment]
Another note for my reference. It looks like the docker client handles forwarding stdin to the container ok, so it is possible. Given the following source `inout.cpp`

```
#include <iostream>
#include <string>

using namespace std;

int main() {
  cout << "start" << endl;
  string input;
  while (cin >> input) {
    cout << "got: " << input << endl;
  }
  cout << "end" << endl;
  return 0;
}
```

and text file

```
hello
world
!
```

executing this command

```sh
$ docker run \
        -i --rm \
        -v $(pwd -P)/inout.cpp:/tmp/dexec/build/inout.cpp \
        dexec/lang-cpp:1.0.2 inout.cpp <test.txt
```

returns

```
start
got: hello
got: world
got: !
end
```

**Comment by andystanton** [comment]
@jayantbh this has been released in v1.0.5

Here is an example of a C program reading from stdin using dexec:

### test.c

```c
#include <stdio.h>

#define BUFFERSIZE 100

int main (int argc, char *argv[])
{
    char buffer[BUFFERSIZE];
    while(fgets(buffer, BUFFERSIZE , stdin)) {
        printf("Read: %s", buffer);
    }
    return 0;
}

```

### test.txt

```txt
hello world!
reading from stdin!
and again!
```

### Command

```sh
$ dexec test.c < test.txt
```

### Output

```
Read: hello world!
Read: reading from stdin!
Read: and again!
```

**Comment by jayantbh** [comment]
Yes, that works!
